pkg/agent/runner.go

```diff
@@ -343,20 +371,56 @@ func (r *Runner) spawnBackgroundWorker(ctx context.Context, logger *zap.Logger,
 	}()
 }
 
+type metricsMgr[M any] struct {
+	// kind is the human-readable name representing this type of metrics.
+	// It's either "system" or "LFC".
+	kind string
+
+	// isActive returns whether these metrics should currently be collected for the VM.
+	//
+	// For example, with LFC metrics, we return false if they are not enabled for the VM.
+	isActive func() bool
+
+	// newMetrics is a callback to update the internal state with new values for these metrics.
+	newMetrics func(metrics M, withLock func())
+}
+
+// metricsPtr wraps core.FromPrometheus with the requirement that the interface be implemented on a
+// pointer type.
+//
+// We use this in getMetricsLoop to construct new values of type M that point to a zero value
+// (rather than being nil pointers themselves).
+type metricsPtr[V any] interface {
```

Comment:
> cc @Omrigan particularly want your feedback on this one. I could go either way on this -- left it as a separate commit so it's easy to revert.

Yeah, I'd revert it. It was educational - I learned `~` is a thing in golang :laughing:  But for real-life usage it'd stick to the trivial approach.

Reply:
Haha fair enough- reverted in 60f835b5fc21bc867d8b3a466f1896a22d44f6dc